src/main/scala/org/kframework/minikore/interfaces/MiniKoreInterface.scala

```diff
@@ -0,0 +1,303 @@
+package org.kframework.minikore.interfaces
+
+object TreeInterface {
+
+  import PatternInterface._
+
+  sealed trait AST
+
+
+  sealed trait Node extends AST with Product {
+    def args: Seq[Pattern]
+
+    def build(children: Seq[Pattern]): Pattern
+  }
+
+  object Node {
+    def unapply(arg: Node): Option[Seq[Pattern]] = Some(arg.args)
+  }
+
+
+  sealed trait Leaf[C] extends AST with Product {
+    def contents: C
+
+    def build(contents: C): Pattern
+  }
+
+  object Leaf {
+    def unapply[C](arg: Leaf[C]): Option[C] = Some(arg.contents)
+  }
+
+
+  sealed trait Leaf2[CC1, CC2] extends Leaf[Product2[CC1, CC2]] with Product2[CC1, CC2] {
+    override def contents: Product2[CC1, CC2] = (_1, _2)
+
+    def apply(_1: CC1, _2: CC2): Pattern
```

Comment:
I am ok with @ehildenb suggestion of renaming that function to `build`. I know that in the end, it's the same, but the non-standard name will suggest to developers that something must be different here.

@ehildenb, why do you say the `apply` methods need to be added back in the implementations?

If you like seeing the `build`ers as `Seq[Pattern] => Pattern`, let's make them that. I'm ok with that. Right now they are actually methods on the `Node`s, which restricts us from ever attaching any extra information to them, e.g., the Kale backend could not be made compatible.

Regarding red/green, which diff has more red than green? I see this one as +912 −127, and 1->2 as +838 −714.